### PR TITLE
Better lock

### PR DESCRIPTION
**Pull request by PhilippvK**
#74

### CODE REVIEW

File: mlonmcu/cli/build.py [hunk 1]
```diff
@@ -114,6 +114,6 @@ def handle(args, ctx=None, require_target=False):
     if ctx:
         _handle(args, ctx, require_target=require_target)
     else:
-        with mlonmcu.context.MlonMcuContext(path=args.home, lock=True) as context:
+        with mlonmcu.context.MlonMcuContext(path=args.home, env_lock="read", latest_session_link_lock=True) as context:
```
Comment (PhilippvK) [inline mlonmcu/cli/build.py:117]: drop `latest_session_link_lock`

Comment (PhilippvK) [inline mlonmcu/cli/build.py:117]: Rename to `deps_lock`, default to `write`

File: mlonmcu/context.py [hunk 2]
```diff
@@ -270,13 +271,20 @@ class MlonMcuContext:
...
 
-    def __init__(self, name: str = None, path: str = None, lock: bool = False):
+    def __init__(self, name: str = None, path: str = None, env_lock: str = "write", latest_session_link_lock=False):
         env_file = resolve_environment_file(name=name, path=path)
         assert env_file is not None, "Unable to find a MLonMCU environment"
         self.environment = UserEnvironment.from_file(env_file)  # TODO: move to __enter__
         setup_logging(self.environment)
-        self.lock = lock
-        self.lockfile = filelock.FileLock(os.path.join(self.environment.home, ".lock"))
+        assert env_lock == "read" or "write"
```
Comment (PhilippvK) [inline mlonmcu/context.py:279]: `assert env_lock in ["read", "write"]`

File: mlonmcu/context.py [hunk 3]
```diff
@@ -287,20 +295,44 @@ def __init__(self, name: str = None, path: str = None, lock: bool = False):
...
-        session_dir = sessions_directory / str(idx)
-        session = Session(idx=idx, label=label, dir=session_dir, config=config)
-        self.sessions.append(session)
-        self.session_idx = idx
-        # TODO: move this to a helper function
-        session_link = sessions_directory / "latest"
-        if os.path.islink(session_link):
-            os.unlink(session_link)
-        os.symlink(session_dir, session_link)
+        if hasattr(self, "latest_session_link_lock"):
```
Comment (PhilippvK) [inline mlonmcu/context.py:298]: Avoid this, just initialize `self.latest_session_link_lock` with `None`

File: mlonmcu/read_write_file_lock/_track [hunk 4]
```diff
@@ -0,0 +1,3 @@
+D7tbI25D3sNmcEllgxAopKp5ly9by1i5:
```
Comment (PhilippvK) [inline mlonmcu/read_write_file_lock/_track:1]: Remove this before merge. (The others as well)

File: mlonmcu/read_write_file_lock/read_write_filelock.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
+#
```
Comment (PhilippvK) [inline mlonmcu/read_write_file_lock/read_write_filelock.py:1]: I would rename the `read_write_lock` directory to something shorter.

Either just `mlonmcu/lock/` or move `mlonmcu/context.py` and `read_write_filelock.py` into a new subdirectory `mlonmcu/context/`

File: mlonmcu/read_write_file_lock/read_write_filelock.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+            if v["type"] == "write" or "read":
+                occupied = True
+                break
+
+        self.lock.release()
+        return occupied
+
+
+if __name__ == "__main__":
+    filepath = "hello"
```
Comment (PhilippvK) [inline mlonmcu/read_write_file_lock/read_write_filelock.py:287]: Move to `test/unit-test/context/test_locks.py`

File: requirements_addition.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# AUTOGENERATED by 'mlonmcu setup -g'
```
Comment (PhilippvK) [inline requirements_addition.txt:1]: Remove file

**Comment by zhouxinyu0723** [comment]
this test case cannot pass. Should I init MlonMcuContext with read lock for this test case?

def test_nest_context(monkeypatch, fake_environment_directory: Path, fake_config_home: Path):
    monkeypatch.chdir(fake_environment_directory)
    create_minimal_environment_yaml(fake_environment_directory / "environment.yml")
    with mlonmcu.context.context.MlonMcuContext() as context:
        assert context
        with mlonmcu.context.context.MlonMcuContext() as context2:
            assert context2

**Comment by PhilippvK** [comment]
@zhouxinyu0723 I think we should refactor this one to check several cases (valid and invalid ones):

- read + read -> success
- read + write -> fail
- write + read -> fail
- write + write -> fail

The default case which is now write + write should fail.

For `success` you can just week using assertions, for `fail` you can use [`pytest.raises`](https://docs.pytest.org/en/stable/reference/reference.html?highlight=raises#pytest.raises) or something similar.

File: mlonmcu/context.py (hunk 3 above)
Comment (zhouxinyu0723) [reply mlonmcu/context.py:298]: Do you mean this in the `__init__()` ?
```
if latest_session_link_lock:
    self.latest_session_link_lock = filelock.FileLock(
        os.path.join(self.environment.home, ".latest_session_link_lock_lock")
    )
else:
    self.latest_session_link_lock = None
```

and change the condition to 

```
if self.latest_session_link_lock:
```

Comment (zhouxinyu0723) [reply mlonmcu/context.py:298]: Or the latest_session_link_lock should always be True.
```
self.latest_session_link_lock = filelock.FileLock(
    os.path.join(self.environment.home, ".latest_session_link_lock_lock")
)
```

and the condition check can be skipped

```

```

Comment (zhouxinyu0723) [reply mlonmcu/context.py]: I chose the later implementation.

Comment (PhilippvK) [reply mlonmcu/context.py]: Perfect!

**Comment by PhilippvK** [comment]
@zhouxinyu0723 Can you give me an update on the status? If this is ready for review, remove the "Draft: " prefix from the PR title.

In addition, to keep the commit history clean, I would recommend doing a rebase on `origin/main` + squashing of similar commits. I can show you how to achieve this, if you need help.

**Comment by PhilippvK** [comment]
Looks good to merge. Thank you! I will update the README later.

**Pull request merged**